rfd/0002-custom-terraform-registry.md

```diff
@@ -107,7 +161,149 @@ We _could_ modify Houston to serve the required files, but as the service is
+described in the [Cloud Artifact Storage Standards](https://github.com/gravitational/cloud/blob/master/rfd/0017-artifact-storage-standards.md)
+RFD.
+
+#### Public naming
+
+I have assumed in this document that the public-facing address for the
+terraform registry host will follow the naming convention for other Teleport
+distribution points, and be `terraform.releases.teleport.dev`.
+
+#### Bucket structure
+
+The S3 bucket backing the registry requires 3 main structures:
+
+1. the **_discovery file_**,
+2. the **_registry_**, and
+3. the **_object store_**
+
+Each structure will live under a separate key prefix, which should allow
+independent policies to be applied to each.
+
+##### 1. The _discovery file_ (key `/.well-known/terraform.json`)
+
+This implements the Terraform Remote Service Discovery Protocol. This is
+essentially a static JSON file that redirects terraform where to the registry
+service on a given host.
+
+For example, this discovery protocol file:
+
+```json
+{
+    "providers.v1": "/registry/"
+}
+```
+
+_**NOTE:**_ The address in the discovery protocol JSON need not be a relative
+path - it could _in theory_ point to a totally different host. It _should_ be
+possible to have the discovery protocol served by `goteleport.com` that would
+direct terraform to `terraform.releases.teleport.dev` for the actual registry,
+if that's the preferred branding.
+
+Note that I have not _personally_ tested this behaviour.
+
+The practical change this would make would be that the provider source would
+change from (in the example above):
+
+```terraform
+source = "terraform.releases.teleport.dev/gravitational/teleport"
+```
+
+to
+
+```terraform
+source = "goteleport.com/gravitational/teleport"
+```
+
+##### 2. The registry (key prefix `/registry`)
+
+This is where the main registry metadata lives. The basic structure is described
+in the [Provider Registry Specification](https://www.terraform.io/internals/provider-registry-protocol),
+but in short there are two parts of the index for each `namespace/provider`
+pair: the `versions` index, and the `download` metadata.
+
+For our purposes, with a namespace of `gravitational` and a provider name of
+`teleport`, the `versions` index file will be located at `/registry/gravitational/teleport/versions`
+
+This is the only object in the repository that need modification as part of
+normal use.
+
+The `download` metadata for each version/OS/architecture triple is stored
+according to the rule
+
+```shell
+/registry/gravitational/teleport/:version/:os/:architecture
+```
+
+Once created, each `download` metadata does not need modification, unless
+a provider object is re-published.
+
+##### 3. The object store (key prefix `/store`)
+
+The object store is where all the release package files are stored. Each
+provider artefact, together with its associated hash & signatures files.
+
+This section of the bucket should be write-only. No objects in this section
+should be modifiable or delete-able.
```

Comment:
Pedantic clarification: `/dev/null` is write only. I think you mean "write once, read many" or the unfortunate acronym WORM.  A humorous site to illustrate the difference: http://www.supersimplestorageservice.com/

Furthermore object versioning on buckets is an all or nothing thing, I don't think we can do WORM on just part of a bucket.

Fortunately, we don't particularly care if artifacts are overwritten when object versioning is on because:

1) we can easily revert to old versions
2) it is logging/SIEM's responsibility to yell at us if this happens.

In some cases (e.g. a release mistake) being able to overwrite (while preserving history) is quite valuable.

Reply:
Wording updated